From 0d3683c8b1c137a8edeea2a6b386deabedfc42ff Mon Sep 17 00:00:00 2001
From: Billy Huang <billyhuang@google.com>
Date: Tue, 11 Mar 2025 10:53:32 -0700
Subject: [PATCH] RESTRICT AUTOMERGE parse userId from authority in opp in the
 same way as ContentProvider

Ensures that userId determination is consistent with ContentProvider

Bug: 395647005
Bug: 395643490
Flag: EXEMPT fix for vulnerability
Test: m com.google.android.bt
Test: atest BluetoothJavaUnitTests:com.android.bluetooth.opp.BluetoothOppSendFileInfoTest
Ignore-AOSP-First: fix for undisclosed vulnerability
(cherry picked from https://googleplex-android-review.googlesource.com/q/commit:73ef4306e91bc500769f2abbad94b61742c43602)
(cherry picked from https://googleplex-android-review.googlesource.com/q/commit:c8c96a43f493b3a497f83af88cde0d2db25177dd)
Merged-In: Iacb9b1fc1fe48d9b8512f6e80bcc6eeb2a5bedbd
Change-Id: Iacb9b1fc1fe48d9b8512f6e80bcc6eeb2a5bedbd
---
 .../opp/BluetoothOppSendFileInfo.java         | 23 ++++++++++++++++++-
 android/app/tests/unit/Android.bp             |  1 +
 .../opp/BluetoothOppSendFileInfoTest.java     | 22 +++++++++++-------
 3 files changed, 37 insertions(+), 9 deletions(-)

diff --git a/android/app/src/com/android/bluetooth/opp/BluetoothOppSendFileInfo.java b/android/app/src/com/android/bluetooth/opp/BluetoothOppSendFileInfo.java
index 7ce134341a..238151f1b1 100644
--- a/android/app/src/com/android/bluetooth/opp/BluetoothOppSendFileInfo.java
+++ b/android/app/src/com/android/bluetooth/opp/BluetoothOppSendFileInfo.java
@@ -262,8 +262,29 @@ public class BluetoothOppSendFileInfo {
         return new BluetoothOppSendFileInfo(fileName, contentType, length, is, 0);
     }
 
+    /**
+     * Determine if the given {@link Uri} is a content uri for another user.
+     *
+     * <p>RFC 2396 s.3.2. states that <tt>'@'</tt> is reserved in the authority component. Its
+     * encoded form should be interpreted as data within the authority component. However,
+     * ContentProvider APIs use the decoded {@link Uri#getAuthority()} with {@link
+     * ContentProvider#getUserIdFromAuthority(String, int)} to determine the <tt>userId</tt> in the
+     * userInfo, rather than {@link Uri#getUserInfo()}. An encoded <tt>'@'</tt>, which is
+     * <tt>'%40'</tt>, is interpreted by ContentProvider as the separator for userInfo and host.
+     *
+     * <p>As an unbundled module, Bluetooth cannot access ContentProvider#getUserIdFromAuthority, so
+     * parse userInfo here from the authority.
+     */
     private static boolean isContentUriForOtherUser(Uri uri) {
-        String uriUserId = uri.getUserInfo();
+        String authority = uri.getAuthority();
+        if (authority == null) {
+            return false;
+        }
+        int atIndex = authority.lastIndexOf('@');
+        if (atIndex == -1) {
+            return false;
+        }
+        String uriUserId = authority.substring(0, atIndex);
         return !TextUtils.isEmpty(uriUserId)
                 && !Objects.equals(uriUserId, String.valueOf(myUserId()));
     }
diff --git a/android/app/tests/unit/Android.bp b/android/app/tests/unit/Android.bp
index 54382cf019..39ab019407 100644
--- a/android/app/tests/unit/Android.bp
+++ b/android/app/tests/unit/Android.bp
@@ -19,6 +19,7 @@ java_defaults {
 
     static_libs: [
         "PlatformProperties",
+        "TestParameterInjector",
         "androidx.media_media",
         "androidx.room_room-migration",
         "androidx.room_room-runtime",
diff --git a/android/app/tests/unit/src/com/android/bluetooth/opp/BluetoothOppSendFileInfoTest.java b/android/app/tests/unit/src/com/android/bluetooth/opp/BluetoothOppSendFileInfoTest.java
index acb58272fb..dcc464d57f 100644
--- a/android/app/tests/unit/src/com/android/bluetooth/opp/BluetoothOppSendFileInfoTest.java
+++ b/android/app/tests/unit/src/com/android/bluetooth/opp/BluetoothOppSendFileInfoTest.java
@@ -35,12 +35,15 @@ import android.net.Uri;
 import android.provider.OpenableColumns;
 
 import androidx.test.platform.app.InstrumentationRegistry;
-import androidx.test.runner.AndroidJUnit4;
 
 import com.android.bluetooth.BluetoothMethodProxy;
 
+import com.google.testing.junit.testparameterinjector.TestParameter;
+import com.google.testing.junit.testparameterinjector.TestParameterInjector;
+
 import org.junit.After;
 import org.junit.Before;
+import org.junit.Ignore;
 import org.junit.Test;
 import org.junit.runner.RunWith;
 import org.mockito.Mock;
@@ -49,7 +52,7 @@ import org.mockito.MockitoAnnotations;
 import java.io.FileInputStream;
 import java.io.IOException;
 
-@RunWith(AndroidJUnit4.class)
+@RunWith(TestParameterInjector.class)
 public class BluetoothOppSendFileInfoTest {
     Context mContext;
     MatrixCursor mCursor;
@@ -122,10 +125,12 @@ public class BluetoothOppSendFileInfoTest {
     }
 
     @Test
-    public void generateFileInfo_withContentUriForOtherUser_returnsSendFileInfoError()
-            throws Exception {
+    @Ignore("TODO: b/377774165 - Use source-built BT APEX instead of prebuilt")
+    public void generateFileInfo_withContentUriForOtherUser_returnsSendFileInfoError(
+            @TestParameter boolean encodedAt) throws Exception {
         String type = "image/jpeg";
-        Uri uri = buildContentUriWithEncodedAuthority((myUserId() + 1) + "@media");
+        String authoritySuffix = encodedAt ? "%40media" : "@media";
+        Uri uri = buildContentUriWithEncodedAuthority((myUserId() + 1) + authoritySuffix);
 
         long fileLength = 1000;
         String fileName = "pic.jpg";
@@ -185,10 +190,11 @@ public class BluetoothOppSendFileInfoTest {
     }
 
     @Test
-    public void generateFileInfo_withContentUriForSameUser_returnsInfoWithCorrectLength()
-            throws Exception {
+    public void generateFileInfo_withContentUriForSameUser_returnsInfoWithCorrectLength(
+            @TestParameter boolean encodedAt) throws Exception {
         String type = "image/jpeg";
-        Uri uri = buildContentUriWithEncodedAuthority(myUserId() + "@media");
+        String authoritySuffix = encodedAt ? "%40media" : "@media";
+        Uri uri = buildContentUriWithEncodedAuthority(myUserId() + authoritySuffix);
 
         long fileLength = 1000;
         String fileName = "pic.jpg";
-- 
2.49.0.1077.gc0e912fd4c-goog

